### PR TITLE
pkg: add tweetnacl crypto library

### PR DESCRIPTION
**Pull request by kaspar030**
I've played a bit with NaCl and wrote a simple example testing NaCl's public key crypto functions some years back.

Turned out that code runs unmodified with tweetnacl [1], which runs unmodified [2] on RIOT.

This PR adds tweetnacl as module and adds a simple public key creation/encryption/decryption example.

My example app uses malloc, which can probably be changed. Also, tweetnacl itself needs a lot of stack. I set (STACKSIZE_DEFAULT*4) for main_thread in the Makefile, but this will very probably fail on some platforms.

[1] http://tweetnacl.cr.yp.to/index.html
[2] I had to fix some signedness warnings. Patch is sent to DJB for upstream inclusion.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Have fun with writing documentation for the header!

*j/k* Nice idea! It's about time to add some stuff on the security site to RIOT.

File: sys/crypto/tweetnacl/LICENSE [hunk 1]
```diff
@@ -0,0 +1,3 @@
+From http://http://tweetnacl.cr.yp.to:
```
Comment (Kijewski) [inline sys/crypto/tweetnacl/LICENSE]: Typo

File: sys/crypto/tweetnacl/randombytes.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+#include <stdint.h>
+
+#include "random.h"
+
+void randombytes(uint8_t *target, uint64_t n)
+{
+    while (n>=4) {
+        uint32_t *pos = (uint32_t*) target;
```
Comment (Kijewski) [inline sys/crypto/tweetnacl/randombytes.c]: I see alignment problems. Better memcpy.

**Comment by Kijewski** [comment]
I did not look into the big file, yet. But this is really promising!

**Comment by kaspar030** [comment]
@Kijewski Regarding alignment, I changed it to use byte-wise copy. memcpy version was >60bytes larger.

**Comment by OlegHahm** [comment]
@kaspar030, what's the state of this one?

**Comment by kaspar030** [comment]
apart from needing a rebase, last time a checked, the example worked fine at least on native an sam r21. That means the package is operational.

**Comment by OlegHahm** [comment]
This one needs a reviewer!

**Comment by OlegHahm** [comment]
@BytesGalore, I had the impression you were looking a bit into security lately. :)

**Comment by OlegHahm** [comment]
We still need a reviewer for that. Any volunteers?

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 3]
```diff
@@ -0,0 +1,810 @@
+#include "tweetnacl.h"
```
Comment (OlegHahm) [inline sys/crypto/tweetnacl/tweetnacl.c]: License is missing

**Comment by OlegHahm** [comment]
Could you please uncrustify the big C file?

File: sys/crypto/tweetnacl/tweetnacl.c (hunk 3 above)
Comment (miri64) [inline sys/crypto/tweetnacl/tweetnacl.c]: https://github.com/RIOT-OS/RIOT/pull/4317/files#diff-0abdd814cde24db625913f773f8bf7e6R3 mentions it's public domain, but yeah a note in the C-file might be helpful, too.

**Comment by BytesGalore** [comment]
I will have a look this afternoon, (sry for the absence)

One thing I directly stumbled upon is a `malloc in the test but no free [1]
[1] https://github.com/RIOT-OS/RIOT/pull/4317/files#diff-e77e412b96be626119405b011444127aR29

**Comment by OlegHahm** [comment]
Good catch: it seems that one could replace this malloc() could be easily replaced by a static array of variable size.

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+#define FOR(i,n) for (i = 0;i < n;++i)
+#define sv static void
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: I would vote to replace such defines with their original keywords,
even if it grow larger than the original 100 tweets NaCl implementation

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+  FOR(i,16) {
+    x[i+16]=a[i];
+    x[i+32]=c[i];
+    x[i+48]=b[i];
+    x[i+64]=d[i];
+  }
+  inv25519(x+32,x+32);
+  M(x+16,x+16,x+32);
+  pack25519(q,x+16);
+  return 0;
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: I guess such returns can be safely removed, when no other value is ever returned by the function

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+}
+
+int crypto_scalarmult_base(u8 *q,const u8 *n)
+{ 
+  return crypto_scalarmult(q,n,_9);
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: same here its always `0`, avoiding such returns would have slimmed the tweetnacl further

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+  0xbb,0x67,0xae,0x85,0x84,0xca,0xa7,0x3b,
+  0x3c,0x6e,0xf3,0x72,0xfe,0x94,0xf8,0x2b,
+  0xa5,0x4f,0xf5,0x3a,0x5f,0x1d,0x36,0xf1,
+  0x51,0x0e,0x52,0x7f,0xad,0xe6,0x82,0xd1,
+  0x9b,0x05,0x68,0x8c,0x2b,0x3e,0x6c,0x1f,
+  0x1f,0x83,0xd9,0xab,0xfb,0x41,0xbd,0x6b,
+  0x5b,0xe0,0xcd,0x19,0x13,0x7e,0x21,0x79
+} ;
+
+int crypto_hash(u8 *out,const u8 *m,u64 n)
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: I'm not against have an alternative, but we already have (documented) `hash` implementations so how about to make this one optional?

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+  x[n] = 128;
+
+  n = 256-128*(n<112);
+  x[n-9] = b >> 61;
+  ts64(x+n-8,b<<3);
+  crypto_hashblocks(h,x,n);
+
+  FOR(i,64) out[i] = h[i];
+
+  return 0;
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: again constant `0` return

File: sys/crypto/tweetnacl/tweetnacl.c [hunk 9]
```diff
@@ -0,0 +1,810 @@
...
+}
+
+sv scalarmult(gf p[4],gf q[4],const u8 *s)
+{
+  int i;
+  set25519(p[0],gf0);
+  set25519(p[1],gf1);
+  set25519(p[2],gf1);
+  set25519(p[3],gf0);
+  for (i = 255;i >= 0;--i) {
```
Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: why not use the `FOR` macro here? 
Or the other way around we could replace the macro with the original call.

Comment (BytesGalore) [inline sys/crypto/tweetnacl/tweetnacl.c]: ok strike this, the macro is forward iterating

File: tests/tweetnacl/main.c [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+int main (void)
+{
+    int res;
+
+    const char message[] = "Test String";
+    struct identity id_alice;
+
+    printf("Creating keypair ALICE...\n");
```
Comment (BytesGalore) [inline tests/tweetnacl/main.c]: `puts` please
and for the others too

File: tests/tweetnacl/main.c [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    unsigned char bob_pk[crypto_box_PUBLICKEYBYTES];
+    unsigned char bob_sk[crypto_box_SECRETKEYBYTES];
+
+    printf("Creating keypair BOB...\n");
+    crypto_box_keypair(bob_pk,bob_sk);
+    printf("Done.\n");
+
+    const unsigned char *m = bytes_to_message(message,strlen(message)+1);
+    unsigned long long mlen = strlen(message)+crypto_box_ZEROBYTES+1;
```
Comment (BytesGalore) [inline tests/tweetnacl/main.c]: I guess `int` would be sufficient

**Comment by BytesGalore** [comment]
Honestly I don't know how to correctly review this PR, since its basically a package but directly integrated like a usual module.
The coding style is obviously non RIOT slimmed to the minimum and the provided documentation is a bit thin ;)

I mean we can focus on the provided test. 
In this respect I would propose to extend it to a unittest that covers all testable functions provided with the module.

**Comment by kaspar030** [comment]
well, yes. I created the module, and then copied a test program I once did for linux, and it worked unchanged for RIOT. So I PR'ed it.

I guess it doesn't make sense to review the upstream code...

-----Original Message-----

**Comment by BytesGalore** [comment]
>well, yes. I created the module, and then copied a test program I once did for linux, and it worked unchanged for RIOT. So I PR'ed it.

Which is totally cool, since it demonstrates the portability towards RIOT :D 

> I guess it doesn't make sense to review the upstream code...

Yeah I would get lost in the abbreviated functions/macros used there.
I tend to propose to use tweetnacl as package rather than a module, since if we integrate it directly to RIOT we should (I feel more like we MUST) at least ensure a certain level of quality regarding our coding conventions and documentation coverage.

**Comment by BytesGalore** [comment]
I made an alternative PR #5687 providing `tweetnacl` as a package

**Comment by kYc0o** [comment]
Feature freeze -> postponed.

**Comment by OlegHahm** [comment]
I would propose the following:
* add the license to the C files
* uncrustify the code
* add some documentation stub, so that this module is at least showing in the documentation somehow

Fine with that, @kaspar030 and @BytesGalore?

**Comment by BytesGalore** [comment]
@OlegHahm :+1:, lets give it a try :)

**Comment by kYc0o** [comment]
@kaspar030 ping. Can you rebase for now? I'll try to make time and test it.

**Comment by miri64** [comment]
Also don't forget about the stuff @OlegHahm was suggesting ;-)

**Comment by miri64** [comment]
Postponed due to Feature Freeze

**Comment by kYc0o** [comment]
I'm testing on OS X and I get a segfault while creating the identity, can you rebase and do the test?

**Comment by kaspar030** [comment]
> I'm testing on OS X and I get a segfault while creating the identity, can you rebase and do the test?

Yeah, there was a bug in randombytes(). That's fixed now.
I also factored the buffers out of main() so we can better measure the stacksize needed by tweetnacl (and removed the "struct id" cruft).

Could you try again?

**Comment by kYc0o** [comment]
Actually I had found the problem and tested it again... (thinking you were maybe busy on other stuff). 

I'll test it asap.

**Comment by kaspar030** [comment]
Should I maybe move the tweetnacl-code into pkg/? I'm very opposed to fiddle with that code. It is half generated upstream by python code, and the API is (more or less) documented publicly. But as it is, it doesn't pass RIOT's coding conventions. Moving the original code to pkg/ would make this clearer...

**Comment by kYc0o** [comment]
It's working now in native OS X. However, it crashes very bad in iotlab-m3. With  my previous patch (which sadly I didn't commit and drop it when you did the changes), but it was basically the same, with the random function adapted from #5687 (which was a package, btw)

**Comment by kaspar030** [comment]
> However, it crashes very bad in iotlab-m3.

Does it crash in randombytes? Works fine on sam21...

**Comment by kYc0o** [comment]
Nope, it crashes while printing `ps()`

**Comment by kYc0o** [comment]
sorry bad button :S

**Comment by kaspar030** [comment]
> Nope, it crashes while printing ps()

Ah, I just realized that the default stacksize for cortex-m is just 1k, so the hack in the makefile is not sufficient. will fix.

**Comment by kYc0o** [comment]
Just did the test in a samr21-xpro and I experience similar problems (trash printed to STDOUT).

**Comment by kaspar030** [comment]
Please try with the latest commit, I adjusted the stacksize for main.

**Comment by kYc0o** [comment]
Great! it works as expected. So would you like to make a pkg? I'm not against it and I agree that the code it's hard to change to RIOT conventions, but it seems that upstream changes are not very often (last change from 2014...).

**Comment by BytesGalore** [comment]
@kYc0o we decided to closed #6072 which provided tweetnacl as package

**Comment by BytesGalore** [comment]
sry, wrong pr number (fixed)

**Comment by kYc0o** [comment]
Yes I know, but @kaspar030 seems to like the idea to move the code to pkg so that's why I asked...

**Comment by miri64** [comment]
I do not exactly understand the argument in #5687 for not including it as a pkg. The only con I see (to many downloads on CI) are basically true for any pkg.

**Comment by kaspar030** [comment]
The idea is to just move the code to pkg/, to make clear that its unmodified upstream code. I just tried, it looks weird to build that way.

**Comment by kYc0o** [comment]
Well, if nobody complains about this "ugly" code for now, I volunteer to make a cleanup soon. Besides that I think this is ready for merging, so please squash and see what Murdock thinks.

**Comment by kYc0o** [comment]
Maybe @OlegHahm [comment](https://github.com/RIOT-OS/RIOT/pull/4317#issuecomment-250320452) could be addressed and this PR can be merged soon?

**Comment by kYc0o** [comment]
If @kaspar030 doesn't have time I can take over this PR, I started to make use of it so it would be nice to have it soon.

**Comment by emmanuelsearch** [comment]
@kYc0o are you suggesting you could take over the PR ;)

**Comment by kaspar030** [comment]
- rebased
- moved code to pkg/

> add the license to the C files

The one from me has a license. for the original file, there's a LICENSE file.

> uncrustify the code

Don't want to modify the upstream code.

> add some documentation stub, so that this module is at least showing in the documentation somehow

did.

**Comment by OlegHahm** [comment]
> Don't want to modify the upstream code.

Sure, that was my proposal in case we were not going for a package.

**Comment by kYc0o** [comment]
Great! Sorry for the pressure :P

**Comment by OlegHahm** [comment]
Hm, can we tag this package somehow so that we can make it a proper package once we "fixed" the packaging system, i.e. introduced some proper caching?

**Comment by kaspar030** [comment]
> Hm, can we tag this package somehow so that we can make it a proper package once we "fixed" the packaging system, i.e. introduced some proper caching?

Do you think that is necessary? If you think "third party code" instead of "package", it doesn't matter if we've actually imported the sources.

**Comment by kYc0o** [comment]
Actually the code is not being downloaded from upstream (and they didn't touch the code since 2014, so I assume it's not maintained anymore) so I don't know if the goal is to put it on some other repo and pull it from there or pull it from upstream (hoping they're not going down)

**Comment by kaspar030** [comment]
> they didn't touch the code since 2014, so I assume it's not maintained
> anymore

With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)
And they'll never add features.

**Comment by kYc0o** [comment]
>With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)
And they'll never add features.

Agreed ;-)

**Comment by kYc0o** [comment]
Code still working properly on native and iotlab-m3. Please squash and let's see what Murdock has to say.

**Comment by kYc0o** [comment]
Ah yes, just please change the commits and PR messages according to the new place on the tree.

**Comment by OlegHahm** [comment]
@kaspar030 wrote:
> >  they didn't touch the code since 2014, so I assume it's not maintained
 anymore

> With crypto libraries, this is considered a good sign. Code was flawless
for two years now. :)

Nice metric for proving security... ;-)

**Comment by OlegHahm** [comment]
> Do you think that is necessary? If you think "third party code" instead of "package", it doesn't matter if we've actually imported the sources.

Maybe it is not necessary, but if we would have proper caching for a package the effect would be almost zero, and just look much nicer.

**Comment by kaspar030** [comment]
> if we would have proper caching for a package the effect would be almost zero, and just look much nicer.

It would mean extra lines in the Makefiles, extra downloads for the cache-cold case, extra complexity, ...

**Comment by OlegHahm** [comment]
It would also mean less code duplication, less ugly code in our beautiful repository, and a consistent style. And I don't see the extra complexity.

However, I do not insist and consider it a waste of time to discuss about such a bagatelle. Nevermind!

File: tests/tweetnacl/main.c [hunk 12]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
```
Comment (kYc0o) [inline tests/tweetnacl/main.c]: Maybe a detail but the doc (if built) will be wrong...

**Comment by kYc0o** [comment]
Some trailing white spaces and a use of malloc found by Murdock... I thought you were able to avoid malloc usage?

**Comment by kYc0o** [comment]
@kaspar030 Can you fix the last Murdock complains? It's almost ready :)

**Comment by kYc0o** [comment]
@kaspar030 ping!

**Comment by emmanuelsearch** [comment]
@kaspar030 seems like a low hanging fruit?

**Comment by kYc0o** [comment]
A fresh Murdock check shows now some cpp check issues... it seems that the headers need "CPP Compatibility" but we agreed on avoiding any modification to the sources, the same for the white spaces.

There's a way to "blacklist" this build for such checks?

For the missing malloc.h it's just a matter of erase that line, it's not needed.

**Comment by OlegHahm** [comment]
@kYc0o, I just checked: the `dist/tools/cppcheck/check.sh` and `dist/tools/externc/check.sh` contain already the possibility to exclude a certain folder - you could add the same for `dist/tools/whitespacecheck/check.sh`.

**Comment by kYc0o** [comment]
@OlegHahm So you propose to make a PR for such modifications before merge this one?

**Comment by OlegHahm** [comment]
Another quickfix would be adding
`pkg/tweetnacl/*/* -diff`
to `.gitattributes`.

**Comment by OlegHahm** [comment]
The following patch should eliminate most of Murdock's complaints:
https://gist.github.com/OlegHahm/b27a743a24213d9bf6e495cc1404cc8b

**Comment by OlegHahm** [comment]
For the record: I don't like this workaround and would still prefer either a proper package or properly imported code.

**Comment by kYc0o** [comment]
I prefer that one, there are several checks failing (cpp, doc, whitespace, etc...) so I think exclude `pkg/tweetnacl/src` and `pkg/tweetnacl/include` would be the best.

**Comment by kYc0o** [comment]
Uhm... downloading the source code from upstream would solve Murdock complains?

**Comment by OlegHahm** [comment]
Yes, external code is not checked by static tests.

**Comment by kYc0o** [comment]
superseded by #6293

**Pull request closed without merging**